test/jdk/java/foreign/TestMemoryCopy.java

```diff
@@ -16,106 +16,390 @@
+        ByteOrder bo = mode.swap ? NON_NATIVE_ORDER : NATIVE_ORDER;
+        //CopyFrom
+        Object srcArr = helper.toArray(base);
+        int srcIndex = mode.direction ? 0 : indexShifts;
+        int srcCopyLen = helper.length(srcArr) - indexShifts;
+        MemorySegment dstSeg = helper.fromArray(srcArr);
+        long dstOffsetBytes = mode.direction ? SEG_OFFSET_BYTES : 0;
+        helper.copyFromArray(srcArr, srcIndex, srcCopyLen, dstSeg, dstOffsetBytes, bo);
+        assertEquals(truth.mismatch(dstSeg), -1);
+        //CopyTo
+        long srcOffsetBytes = mode.direction ? 0 : SEG_OFFSET_BYTES;
+        Object dstArr = helper.toArray(base);
+        MemorySegment srcSeg = helper.fromArray(dstArr);
+        int dstIndex = mode.direction ? indexShifts : 0;
+        int dstCopyLen = helper.length(dstArr) - indexShifts;
+        helper.copyToArray(srcSeg, srcOffsetBytes, dstArr, dstIndex, dstCopyLen, bo);
+        MemorySegment result = helper.fromArray(dstArr);
+        assertEquals(truth.mismatch(result), -1);
+    }
+
+    @Test(dataProvider = "copyModesAndHelpers")
+    public void testUnalignedCopy(CopyMode mode, CopyHelper<Object> helper, String helperDebugString) {
+        int bytesPerElement = (int)helper.elementLayout.byteSize();
+        int indexShifts = SEG_OFFSET_BYTES / bytesPerElement;
+        MemorySegment base = srcSegment(SEG_LENGTH_BYTES);
+        ByteOrder bo = mode.swap ? NON_NATIVE_ORDER : NATIVE_ORDER;
+        //CopyFrom
+        Object srcArr = helper.toArray(base);
+        int srcIndex = mode.direction ? 0 : indexShifts;
+        int srcCopyLen = helper.length(srcArr) - indexShifts;
+        MemorySegment dstSeg = helper.fromArray(srcArr);
+        long dstOffsetBytes = mode.direction ? (SEG_OFFSET_BYTES - 1) : 0;
+        helper.copyFromArray(srcArr, srcIndex, srcCopyLen, dstSeg, dstOffsetBytes, bo);
+        //CopyTo
+        long srcOffsetBytes = mode.direction ? 0 : (SEG_OFFSET_BYTES - 1);
+        Object dstArr = helper.toArray(base);
+        MemorySegment srcSeg = helper.fromArray(dstArr);
+        int dstIndex = mode.direction ? indexShifts : 0;
+        int dstCopyLen = helper.length(dstArr) - indexShifts;
+        helper.copyToArray(srcSeg, srcOffsetBytes, dstArr, dstIndex, dstCopyLen, bo);
+    }
+
+    /***** Utilities *****/
 
-    @Test(dataProvider = "slices")
-    public void testCopy(SegmentSlice s1, SegmentSlice s2) {
-        int size = Math.min(s1.size(), s2.size());
-        //prepare source and target segments
-        for (int i = 0 ; i < size ; i++) {
-            BYTE_HANDLE.set(s2.segment.asSlice(i), (byte)0);
+    public static MemorySegment srcSegment(int bytesLength) {
+        byte[] arr = new byte[bytesLength];
+        for (int i = 0; i < arr.length; i++) {
+            arr[i] = (byte)i;
         }
-        for (int i = 0 ; i < size ; i++) {
-            BYTE_HANDLE.set(s1.segment.asSlice(i), (byte) i);
+        return MemorySegment.ofArray(arr);
```

Comment:
This segment should be made readonly, to assert all checkAccess() calls.

Reply:
Ok - you added this to slice, e.g. here. I have overseen this! https://github.com/openjdk/panama-foreign/pull/555/files#diff-cf69e75a9e792b2f93dec150609654fe2dc3b4f3b45a8fbdca1b5b450d52173bR82